sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakeFileAsyncClient.java

```diff
@@ -148,6 +183,368 @@ public String getFileName() {
+     *
+     * @param data The data to write to the file. Unlike other upload methods, this method does not require that the
+     * {@code Flux} be replayable. In other words, it does not have to support multiple subscribers and is not expected
+     * to produce the same values across subscriptions.
+     * @param length The exact length of the data. It is important that this value match precisely the length of the
+     * data emitted by the {@code Flux}.
+     * @param parallelTransferOptions {@link ParallelTransferOptions} used to configure buffered uploading.
+     * @param headers {@link PathHttpHeaders}
+     * @param metadata Metadata to associate with the resource.
+     * @param requestConditions {@link DataLakeRequestConditions}
+     * @return A reactive response containing the information of the uploaded file.
+     */
+    public Mono<Response<PathInfo>> uploadWithResponse(Flux<ByteBuffer> data, long length,
+        ParallelTransferOptions parallelTransferOptions, PathHttpHeaders headers, Map<String, String> metadata,
+        DataLakeRequestConditions requestConditions) {
+        try {
+            Objects.requireNonNull(data, "'data' must not be null");
+            DataLakeRequestConditions validatedRequestConditions = requestConditions == null
+                ? new DataLakeRequestConditions() : requestConditions;
+            /* Since we are creating a file with the request conditions, everything but lease id becomes invalid
+             after creation, so remove them for the append/flush calls. */
+            DataLakeRequestConditions validatedUploadRequestConditions = new DataLakeRequestConditions()
+                .setLeaseId(validatedRequestConditions.getLeaseId());
+            final ParallelTransferOptions validatedParallelTransferOptions =
+                ModelHelper.populateAndApplyDefaults(parallelTransferOptions);
+            long fileOffset = 0;
+
+            Function<Flux<ByteBuffer>, Mono<Response<PathInfo>>> uploadInChunksFunction = (stream) ->
+                uploadInChunks(stream, fileOffset, length, validatedParallelTransferOptions, headers,
+                    validatedUploadRequestConditions);
+
+            BiFunction<Flux<ByteBuffer>, Long, Mono<Response<PathInfo>>> uploadFullMethod =
+                (stream, lengthUploaded) -> uploadWithResponse(ProgressReporter
+                        .addProgressReporting(stream, validatedParallelTransferOptions.getProgressReceiver()),
+                    fileOffset, length, headers, validatedUploadRequestConditions);
+
+            return createWithResponse(null, null, headers, metadata, requestConditions)
+                .then(UploadUtils.determineUploadFullOrChunked(data, validatedParallelTransferOptions,
+                uploadInChunksFunction, uploadFullMethod));
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
```

Comment:
I wonder if it'd be helpful to have an abstract type we define like Uploader that has abstract methods for `validateOptions`, `initialize`, `determineUploadFullOrChunked`, `uploadInChunks`, `uploadFull`, and `finalize` and then it's like execute/upload/whatever method strings these together and each package has a subtype that overrides each of these methods? Does that make sense and/or do you see any value in that? On the one hand, it seems cleaner to me to have this logic centralized and to deal with well defined overload. On the other hand, I can see why that would feel stiff or overly structured.

Reply:
https://github.com/Azure/azure-sdk-for-java/issues/8383